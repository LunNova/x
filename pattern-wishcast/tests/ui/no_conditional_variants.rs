// SPDX-FileCopyrightText: 2025 LunNova
//
// SPDX-License-Identifier: MIT

//! Test that we get a helpful error when no conditional variants exist

use pattern_wishcast::pattern_wishcast;

pattern_wishcast! {
	enum Value is <P: PatternFields> = {
		HostValue { value: String },
		TupleValue { elements: Vec<Self> },
	};

	type FlexValue = Value is _;
	type StrictValue = Value is HostValue(_) | TupleValue(_);

	#[derive(SubtypingRelation(upcast=to_flex, downcast=try_to_strict))]
	impl StrictValue : FlexValue;
}

fn main() {}
