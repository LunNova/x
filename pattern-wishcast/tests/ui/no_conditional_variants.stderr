error: No conditional variants found for enum `Value`. All variants are included in all pattern types, making them identical. Either: 1) Add variants that are excluded from some pattern types, 2) Use a single type alias instead of multiple identical ones, or 3) Remove `is <P: PatternFields>` if you don't need strictness patterns.
  --> tests/ui/no_conditional_variants.rs:11:1
   |
11 | / pattern_wishcast! {
12 | |     enum Value is <P: PatternFields> = {
13 | |         HostValue { value: String },
14 | |         TupleValue { elements: Vec<Self> },
...  |
21 | |     impl StrictValue : FlexValue;
22 | | }
   | |_^
   |
   = note: this error originates in the macro `pattern_wishcast` (in Nightly builds, run with -Z macro-backtrace for more info)
